sdk/core/core-http/src/credentials/accessTokenRefresher.ts

```diff
@@ -0,0 +1,67 @@
+import { AccessToken, TokenCredential, GetTokenOptions } from "@azure/core-auth";
+
+/**
+ * Helps the core-http token authentication policies with requesting a new token if we're not currently waiting for a new token.
+ */
+export class AccessTokenRefresher {
+  private promise: Promise<AccessToken | undefined> | undefined;
+  private lastCalled: number | undefined;
+
+  constructor(
+    private credential: TokenCredential,
+    private scopes: string | string[],
+    private requiredMillisecondsBeforeNewRefresh: number
+  ) {}
+
+  private readyForNewRefresh() {
+    // We're only ready for a new refresh if the required milliseconds have passed.
+    return (
+      !this.lastCalled || Date.now() - this.lastCalled > this.requiredMillisecondsBeforeNewRefresh
+    );
+  }
+
+  /**
+   * Stores the time in which it is called,
+   * then requests a new token,
+   * then sets this.promise to undefined,
+   * then returns the token.
+   * @param options getToken options
+   */
+  private async getToken(options: GetTokenOptions): Promise<AccessToken | undefined> {
+    this.lastCalled = Date.now();
+    const token = (await this.credential.getToken(this.scopes, options)) || undefined;
```

Comment:
I'm not sure I understand the need for `|| undefined` - isn't `getToken` going to return a truthy value or undefined already?

Reply:
This is inherited from: https://github.com/Azure/azure-sdk-for-js/pull/10085/files#diff-65d593c4170e3a543713c89a60a23baaL87 but I would need to dig more to give a better answer.

Reply:
`getToken` returns `null` if it doesn't have anything to return, so perhaps you could use `null` instead of `undefined` to pass through that behavior.

Reply:
getToken has to return either the token or `undefined`, but I can move the `|| undefined` to the return of the function! That will make it better. I'll do that :)